include/seastar/core/internal/io_sink.hh

```diff
@@ -21,7 +21,7 @@
 
 #pragma once
 
-#include <seastar/core/circular_buffer.hh>
+#include "seastar/core/chunked_fifo.hh"
 #include <seastar/core/internal/io_request.hh>
```

Comment:
> This is to avoid the large contiguous allocation done by
> circular_buffer when there many pending IOs, as such allocations can
> fail on a long running system due to fragmentation.

How many pending ios do you have?


ScyllaDB works hard to ensure a 128kB allocation is "easy". 128kB/32 = 4096, hard to see how a single shard should have 4096 concurrent I/Os with any reasonable latency. Maybe your contiguous allocation threshold is lower?

Reply:
As a side note, with the uring backend network IO also goes through the io_sink so this becomes more likely to happen.

Reply:
That's a good point. Still, I'd like to understand the context.

Reply:
btw, chunked_fifo could be optimized to bump items_per_chunk upward so that sizeof(chunk) approaches a power of two, to avoid wasted space.

Reply:
chunked_fifo is currently written to assume that items_per_chunk is a power of two, so sizeof(chunk) will be a power of two if and only if sizeof(T) is a power of two.

We could allow items_per_chunk be non-power-of-two, at the cost of slightly more expensive (but probably negligible) index calculation.

Reply:
I think it can be brought down to zero, we don't have random access.

Reply:
Thanks @avikivity for your comments.

> Maybe your contiguous allocation threshold is lower?

We also use 128K as our maximum "allowed" allocation size.

The number of IOs was in the range 4000-8000, enough to cause a ~450K allocation in this code. Evidently to exceed 128K would be a smaller number but still more than 1,000.

> hard to see how a single shard should have 4096 concurrent I/Os with any reasonable latency. 

I do not think the latency is reasonable at this point. The specific issue I'm thinking of occurs in a stress test so there are many outstanding IOs. 

Note also that "large IO backlog" does not necessary mean poor latency for all requests. E.g., there may be a bulk background operation creating lots of IO, but at the same time latency sensitive operations which arrive randomly and the IO priority stuff should allow the latency sensitive stuff to jump the queue. However, your point stands for the specific case of `io_sink` since this is a single per-shard queue which is after the scheduling decision has been made. So a large io_sink does imply poor latency for all requests.

The strongest argument to "allowing" may IOs though is that we don't entirely control the disk behavior. Even if we have io properties configured property and play by the rules, there is nothing stopping the disk from changing its performance characteristics suddenly for 100s of millseconds. Such events are fairly common for both attached EBS disks and local SSDs on the usual instance types in EC2 (this is just the cloud I know the best, I don't think it's wildly different anywhere else).

In that case, when we are doing 3,000 IO/s on one shard it only takes about 300 ms to accumulate > 1000 IOs in flight, and I don't think the current IO scheduler will prevent it.

Reply:
@nyh wrote:

> We could allow items_per_chunk be non-power-of-two, at the cost of slightly more expensive (but probably negligible) index calculation.

On the Redpanda side we have `chunked_vector` which offers something like `std::vector`, including `operator[]` and we decided to use power-of-two size in bytes rather than power-of-two size in elements, to save memory at the cost of more expensive indexing. The indexing cost for the constant-divisor `/` is not all that negligible though and we do see regressions in some cases where `std::vector` was replaced with this class. In many cases we just have to switch to iteration rather than `operator[]` indexing as iteration is faster.

A useful API is an inverted iterator like `container::visit_all(Fn& f)` where you pass in the visitor and the container calls it for every element. This can ~remove the cost of the more complicated iterators since it can do a simple iteration over the entire chunk.

Reply:
> The strongest argument to "allowing" may IOs though is that we don't entirely control the disk behavior. Even if we have io properties configured property and play by the rules, there is nothing stopping the disk from changing its performance characteristics suddenly for 100s of millseconds. Such events are fairly common for both attached EBS disks and local SSDs on the usual instance types in EC2 (this is just the cloud I know the best, I don't think it's wildly different anywhere else).

Why does disk behavior matter? The rate at which IO scheduler dispatches requests is only controlled by the io-properties file, it doesn't care if disk completes them at some different rate (in fact, it sometimes cares if disk completes them slower than expected, but it's short-term and can only slow down the dispatch rate)

Reply:
> The number of IOs was in the range 4000-8000

With default io-latency-goal of 0.75ms? That's about 10M operations/sec for a single shard

Reply:
> Thanks @avikivity for your comments.
> 
> > Maybe your contiguous allocation threshold is lower?
> 
> We also use 128K as our maximum "allowed" allocation size.
> 
> The number of IOs was in the range 4000-8000, enough to cause a ~450K allocation in this code. Evidently to exceed 128K would be a smaller number but still more than 1,000.
> 
> > hard to see how a single shard should have 4096 concurrent I/Os with any reasonable latency.
> 
> I do not think the latency is reasonable at this point. The specific issue I'm thinking of occurs in a stress test so there are many outstanding IOs.
> 
> Note also that "large IO backlog" does not necessary mean poor latency for all requests. E.g., there may be a bulk background operation creating lots of IO, but at the same time latency sensitive operations which arrive randomly and the IO priority stuff should allow the latency sensitive stuff to jump the queue. However, your point stands for the specific case of `io_sink` since this is a single per-shard queue which is after the scheduling decision has been made. So a large io_sink does imply poor latency for all requests.
> 
> The strongest argument to "allowing" may IOs though is that we don't entirely control the disk behavior. Even if we have io properties configured property and play by the rules, there is nothing stopping the disk from changing its performance characteristics suddenly for 100s of millseconds. Such events are fairly common for both attached EBS disks and local SSDs on the usual instance types in EC2 (this is just the cloud I know the best, I don't think it's wildly different anywhere else).

The I/O scheduler will globally issue as many ops as it thinks the disk needs to reach --io-latency-goal. If you have one shard and you issue 1000 ops, then on a 100 usec disk it means your latency goal was 100 ms.


> 
> In that case, when we are doing 3,000 IO/s on one shard it only takes about 300 ms to accumulate > 1000 IOs in flight, and I don't think the current IO scheduler will prevent it.

It should.

> 
> A latency regression is not avoidable when the disks do that, but this much preferable to crashing due to a large allocation.

I don't disagree, but what you describe should not have happened.

Reply:
> Why does disk behavior matter? The rate at which IO scheduler dispatches requests is only controlled by the io-properties file, it doesn't care if disk completes them at some different rate (in fact, it sometimes cares if disk completes them slower than expected, but it's short-term and can only slow down the dispatch rate)

Sorry this was confusion on my end, I was mixing together this issue and #70 which is related to the number of the requests in the disk and so disk completion rate matters. You should mostly ignore anything related in my reply.

Reply:
> With default io-latency-goal of 0.75ms? That's about 10M operations/sec for a single shard

Sorry I mean the number of IOs in the `io_sink` was that amount, which is just reverse engineered from the large allocation: i.e., the `_pending_ios` buffer was 450K so we can conclude there are 4000-8000 IOs in it.

Yes, we are using the default io-latency goal.

Reply:
And what about io_properties.yaml? Maybe it happens without it, so scheduler is not limited in the amount of requests it can dispatch in one tick?

Reply:
@xemul wrote:

> And what about io_properties.yaml? Maybe it happens without it, so scheduler is not limited in the amount of requests it can dispatch in one tick?

This case happens without any specific io properties set, yes.

However, though the scheduler may limit the size of the io_sink in _some_ cases (and it may in fact do it in this case), I don't think it's a guarantee:

 - Some operations bypass the scheduler/io-queue and just jump directly to the io_sink, like `fsync`. In Redpanda we often have a 1:1 ratio of write to fsync, so there may be a lot of fsync in flight.
 - With io_uring, additional operations end up in the io_sink, which also bypass the IO scheduler, such as socket operations.
 - There is no guarantee that all operations are flushed out of io_sink every tick: e.g., we have only 1K iocbs to absorb requests coming out of the io_sink, if those fill up the io_sink will carry over operations to the next tick.
 
I think all three of these were pointed out to me by @StephanDollberg.

@xemul wrote:

> Why does disk behavior matter? The rate at which IO scheduler dispatches requests is only controlled by the io-properties file, it doesn't care if disk completes them at some different rate

You are right, I was in part mixing this up with a reactor stall issue (tags exhausted at block layer) which does rely on IO completion. This _mostly_ doesn't rely on io completion behavior (though it does a bit per the last item in the list above: the io_sink may carry over ops from tick-to-tick depending on iocb exhaustion and that relates to IO completion, since it is completion that frees iocbs).

@avikivity wrote:

> The I/O scheduler will globally issue as many ops as it thinks the disk needs to reach --io-latency-goal. If you have one shard and you issue 1000 ops, then on a 100 usec disk it means your latency goal was 100 ms.

Right, and we were not using io properties here (we are changing that), though per above it I don't think it is a comprehensive solution.

Reply:
> @xemul wrote:
> 
> > And what about io_properties.yaml? Maybe it happens without it, so scheduler is not limited in the amount of requests it can dispatch in one tick?
> 
> This case happens without any specific io properties set, yes.
> 
> However, though the scheduler may limit the size of the io_sink in _some_ cases (and it may in fact do it in this case, if a suitable io properties were set), I don't think it's a guarantee:
> 
> * Some operations bypass the scheduler/io-queue and just jump directly to the io_sink, like `fsync`. In Redpanda we often have a 1:1 ratio of write to fsync, so there may be a lot of fsync in flight.

You should switch to O_DSYNC. See the ScyllaDB commitlog code, and all our internal discussions around it, since it's quite tricky. On disks with queue/write_cache = write through it's much much much better.



> * With io_uring, additional operations end up in the io_sink, which also bypass the IO scheduler, such as socket operations.

I think this is compelling.

> * There is no guarantee that all operations are flushed out of io_sink every tick: e.g., we have only 1K iocbs to absorb requests coming out of the io_sink, if those fill up the io_sink will carry over operations to the next tick.

That's an abnormal situation that should be fixed. You're adding 0.5ms latency (or maybe more) because we didn't reserve enough buffer space.

Perhaps we should preempt (add_urgent_task) if _pending_tasks grows beyond a reasonable number (100). Then we never reach those crazy counts. 


> 
> I think all three of these were pointed out to me by @StephanDollberg.
> 
> @xemul wrote:
> 
> > Why does disk behavior matter? The rate at which IO scheduler dispatches requests is only controlled by the io-properties file, it doesn't care if disk completes them at some different rate
> 
> You are right, I was in part mixing this up with a reactor stall issue (tags exhausted at block layer) which does rely on IO completion. This _mostly_ doesn't rely on io completion behavior (though it does a bit per the last item in the list above: the io_sink may carry over ops from tick-to-tick depending on iocb exhaustion and that relates to IO completion, since it is completion that frees iocbs).
> 
> @avikivity wrote:
> 
> > The I/O scheduler will globally issue as many ops as it thinks the disk needs to reach --io-latency-goal. If you have one shard and you issue 1000 ops, then on a 100 usec disk it means your latency goal was 100 ms.
> 
> Right, and we were not using io properties here (we are changing that), though per above it I don't think it is a comprehensive solution.

I believe the correct solution is to schedule a reactor poll when the queue is large enough. The reason for queuing in the first place is to amortize the system call / cache thrashing / etc over many requests, but it we already have many requests, let's flush the queue.

Reply:
> That's an abnormal situation that should be fixed. You're adding 0.5ms latency (or maybe more) because we didn't reserve enough buffer space.
> 
> Perhaps we should preempt (add_urgent_task) if _pending_tasks grows beyond a reasonable number (100). Then we never reach those crazy counts.

I think it is abnormal, yes. We only see this the large allocation in quite specific cases and while we have actually seen the 1K iocbs exhaustion too, it is again a special case, often transient. I'm just trying to be somewhat exhaustive in the cases where this can occur currently.

Reply:
> I believe the correct solution is to schedule a reactor poll when the queue is large enough. The reason for queuing in the first place is to amortize the system call / cache thrashing / etc over many requests, but it we already have many requests, let's flush the queue.

This seems reasonable to me, but it's a much larger change in terms of behavior and is optimizing what I think you are arguing is an edge case. Also, even if we do signal preemption in this case, it doesn't really guarantee anything about the size of the io queue/sink because an application thread may still blast a lot of stuff into the queue without suspending, or do large reads/write which get split up by the io queue/sink to conform to the max io size which also results in many ops.

See also my attempt to [bypass a lot of this discussion here](https://github.com/scylladb/seastar/pull/2357#issuecomment-2263754754).